packages/payment-detection/src/reference-based-detector.ts

```diff
@@ -0,0 +1,224 @@
+import { ExtensionTypes, PaymentTypes, RequestLogicTypes } from '@requestnetwork/types';
+import Utils from '@requestnetwork/utils';
+import getBalanceErrorObject from './balance-error';
+import PaymentReferenceCalculator from './payment-reference-calculator';
+
+import { BigNumber } from 'ethers';
+import { IReferenceBased } from 'types/src/extensions/pn-any-reference-based-types';
+
+/**
+ * Abstract class to extend to get the payment balance of reference based requests
+ */
+export default abstract class ReferenceBasedDetector<TPaymentEventParameters>
```

Comment:
This abstract class was created as the refactoring for code re-usability, based on input-data ETH payment "detector" (the classes that did not have specif names until now, and which are not the retriever). The logic is kept untouched compared to ETH, except for a few factorizations.